### ISSUE TITLE
Proposal for improved show methods for structured matrices

### ISSUE THREAD

**Issue by gajomi** [JuliaLang/julia#13671]
I would like to suggest that matrix types with structural zeros make this structure evident in the default printing. For example, the lower triangulat elements of UpperTriangular instances are guaranteed to be zero so that the current printing of this
```julia
>>>A = UpperTriangular(im*rand(4,4))
4x4 UpperTriangular{Complex{Float64},Array{Complex{Float64},2}}:
 0.0+0.895247im  0.0+0.7869im    0.0+0.280563im  0.0+0.228014im
 0.0+0.0im       0.0+0.835397im  0.0+0.109216im  0.0+0.843416im
 0.0+0.0im       0.0+0.0im       0.0+0.366928im  0.0+0.251685im
 0.0+0.0im       0.0+0.0im       0.0+0.0im       0.0+0.658024im
```
would under my proposal become
```julia
>>>A = UpperTriangular(im*rand(4,4))
4x4 UpperTriangular{Complex{Float64},Array{Complex{Float64},2}}:
 0.0+0.895247im  0.0+0.7869im    0.0+0.280563im  0.0+0.228014im
                 0.0+0.835397im  0.0+0.109216im  0.0+0.843416im
                                 0.0+0.366928im  0.0+0.251685im
                                                 0.0+0.658024im
```

with similar behavior for lower LowerTriangular,Diagonal, BiDiagonal, et cetera. Arbitrary sparse matrices would keep their current behavior. Only the structural zeros would be whited out (e.g. the zero valued elements of a Diagonal type would still print as zero). The dense array versions of these would still have all their zeros explicitly printed. Furthermore it would be nice to see a standard way of achieving this for user defined AbstractMatrix types with structural zeros (I was partially able to do so by defining custom ``print_matrix_repr`` methods for my types, but this did not give me the full behavior above; see comments below).

I think this is a nice behavior, and have been using it extensively while work on a project with structured block matrices, in which I can immediately discern the structure of blocks in a matrix. Obviously there is quite a bit of subjectivity here though. However, there is quite a bit of precedent for this in the literature, where one sees this implicit zero printing all the time (citation needed, but alas, I loaned my Strang/Trefthen references years ago; I'm pretty sure they do this though). So, even if this may not be the preferred default behavior, it would be nice to have it as an option.

**Comment by jiahao** [comment]
Can you show us your custom `print_matrix_repr` methods that implement this particular form of matrix printout?

**Comment by tkelman** [comment]
I like the idea. Diagonal could be printed just as the vector, but this works too.

**Comment by malmaud** [comment]
+1; would love this.

**Comment by gajomi** [comment]
@jiahao - Sorry, my reference `print_matrix_repr` was a rather incoherent glimpse into my debugging effort, wherein I could emulate the compact printing of elements by replacing
```julia
compact ? showcompact_lim(io, el) : show(io, el)
```
with
```julia
Base.showcompact(io, " "^length(sprint(Base.showcompact, el)))
```
in the definition [show.jl:1185](https://github.com/JuliaLang/julia/tree/0ff703b40afddf9b705bd6a06d3a59cb4c089ea5/base/show.jl#L1185). But, this doesn't give you, for example, the correct alignment which is still a mystery for me. I don't have a way of getting the printing above uses the regular display methods. 

The actual way that I get the output above is much more embarrassing: by redirecting io...
```julia
function show(X::UpperTriangular)
    buf = IOBuffer()
    Base.showlimited(buf,X)
    s = takebuf_string(buf)
    print(replace(s,"0.0+0.0im"," "^9))
end
```

I then have to call show. Which will basically work as long as the caveats I mentioned about incidental zeros in the structured part don't appear. So yeah, the whole reason I was using complex elements was so that I could use this naughty hack to move my project along. Obviously this is not how this should actually be done. 

Anyway, I edited the initial comment to reflect this.

**Comment by gajomi** [comment]
I suppose I should also say that this was in an IJulia notebook (perhaps other repls would behavior differently).

**Comment by gajomi** [comment]
@tkelman-
Yeah, I think just a vector would be good for viewing Diagonals in isolation. Perhaps staggered vectors would work for BiDiagonals/TriDiagonals (although one would have to indicate upper versus lower somehow). For my application with blocks, however, I really do want the full matrix view, since the diagonal is part of a larger structure. So maybe it would be nice to have an api to switch between different ways of showing these.

**Comment by gajomi** [comment]
OK, since there seem to be some interest, I parted with the hacks above and made an effort to try to draft an implementation strategy aligned with the proposal. This seems to be working for me on 0.4.0. The core of the strategy is to supplement the current [Base.print_matrix_repr](https://github.com/JuliaLang/julia/tree/0ff703b40afddf9b705bd6a06d3a59cb4c089ea5/base/show.jl#L1016) with an additional method
```julia
function Base.print_matrix_row(io::IO,X::AbstractMatrix, A::Vector,
    i::Integer, cols::AbstractVector, sep::AbstractString)
    for k = 1:length(A)
        j = cols[k]
        if isassigned(X,i,j)
            x = X[i,j]
            a = Base.alignment(x)
            sx = sprint(Base.showcompact_lim, x)
        else
            a = Base.undef_ref_alignment
            sx = Base.undef_ref_str
        end
        l = repeat(" ", A[k][1]-a[1])
        r = repeat(" ", A[k][2]-a[2])
        isstructuralzero(X,i,j) ? print(io, l, " "^length(sx), r) : print(io, l, sx, r)
        if k < length(A); print(io, sep); end
    end
end
```
which specializes dispatch to AbstractMatrix types and where I have introduced a change in the second to last line that calls a proposed Base method `isstructuralzero` that returns true if the indices correspond to structural zeros.  This line tries to keep the printed form of structure matrices close to the original form, where the zeros string are replaced with whitespaces of the same length. With this change one must then define the methods for the relevant types:
```julia
isstructuralzero(X::AbstractMatrix,i::Integer,j::Integer) = false
isstructuralzero(X::Diagonal,i::Integer,j::Integer) = i!=j
isstructuralzero(X::Tridiagonal,i::Integer,j::Integer) = !(i==j-1||i==j||i==j+1)
function isstructuralzero(X::Bidiagonal,i::Integer,j::Integer)
    X.isupper ? !(i==j || i==j-1) : !(i==j || i==j+1)
end
isstructuralzero(X::UpperTriangular,i::Integer,j::Integer) = i>j
isstructuralzero(X::LowerTriangular,i::Integer,j::Integer) = i<j
```
after which point it is possible to write, for example
```julia
mats = Any[Diagonal(rand(4)),
    Tridiagonal(rand(3),rand(4),rand(3)),Bidiagonal(rand(4),rand(3),true),
    UpperTriangular(rand(4,4)),LowerTriangular(rand(4,4))]
for mat in mats
    display(mat)
end
```
which gives (for me, within an IJulia notebook)
```julia
4x4 Diagonal{Float64}:
 0.395004                              
           0.453257                    
                     0.258339          
                               0.973754
4x4 Tridiagonal{Float64}:
 0.631123  0.753763                     
 0.107024  0.422282  0.427519           
           0.892995  0.840954   0.136196
                     0.0933635  0.938127
4x4 Bidiagonal{Float64}:
 0.871382  0.131416                    
           0.302109  0.717238          
                     0.642198  0.200963
                               0.945944
4x4 UpperTriangular{Float64,Array{Float64,2}}:
 0.795011  0.329098  0.582715  0.0108027
           0.975062  0.272005  0.59333  
                     0.266454  0.296902 
                               0.0232241
4x4 LowerTriangular{Float64,Array{Float64,2}}:
 0.197951                               
 0.860549   0.962452                    
 0.0153369  0.669759  0.946966          
 0.50451    0.747806  0.160348  0.824775
```
I have also tried this out with large matrices where the row and columns don't fit, and it looks good to me (although I think there is certainly room for optimization).

Anyway, with this strategy users can enable structural zero printing on their own types by defining an associated `isstructuralzero` method. For example, with the Frobenius matrices in [SpecialMatrices.jl](https://github.com/jiahao/SpecialMatrices.jl) we can write
```julia
using SpecialMatrices

isstructuralzero(X::Frobenius,i::Integer,j::Integer) = !(i==j || j==X.colidx && j<i)
```
to get
```julia
>>>F=Frobenius(3, collect(1.0:3.0))
6x6 SpecialMatrices.Frobenius{Float64}:
 1.0                         
      1.0                    
           1.0               
           1.0  1.0          
           2.0       1.0     
           3.0            1.0
```

I don't have a good sense for how this approach generalizes, but it was simple to implement. I haven;t tried it out, but I imagine there must be a generalization to higher dimensional arrays, if there was use for this.

**Comment by tkelman** [comment]
I like the sound of everything I'm hearing. Can you put the code in a branch and open as a pull request?

**Comment by andreasnoack** [comment]
Indeed. This looks promising.

**Comment by artkuo** [comment]
I have recent experience with `print_matrix_row` and this looks solid. One tiny nitpick: Rather than use a ternary with `isstructuralzero`, why not use `if`? This isn't an assignment, and I personally prefer explicit `if` rather than the more concise (but in my opinion less clear) ternary. This is admittedly a python-esque thing, better explicit than implicit, but does Julia have a similar philosophy? That said, nice solution.

**Comment by gajomi** [comment]
It took forever to build Julia on my janky old laptop (I had been working with the binaries before), but everything seems fine. Here is the [pull request](https://github.com/JuliaLang/julia/pull/13732). 

Maybe I'm feeling a bit greedy today though. I think it could also be cool if we got similar printing for relevant factorization types. For example svds/eigs would have pretty printed diagonals, LU factorization would look like adjacent triangles and so on and so forth. It seem that at present thees types of structured factors are represented as dense matrices, so that the scheme above doesn't automatically extend to the factor printing. I guess there must be a reason for this?

**Comment by artkuo** [comment]
Great ideas. I'm not sure why those factorizations don't take advantage of the structured types. There is even an (unexported) `UnitLowerTriangular` type that could go with LU, so the L could even be printed with pure ones on the diagonal. Also for some reason `eye` is dense and nowhere as cool as `I`.

**Comment by andreasnoack** [comment]
See https://github.com/JuliaLang/julia/pull/12105 and the issue linked there. I think I've changed my mind since I opened that PR and now I agree with @StefanKarpinski that we should print the factorizations as the matrix it represents. However, I've realized that we cannot reconstruct the underlying matrices from a generalized non-symmetric eigenfactorization sο we have to adjust that factorization before this work can be completed, but there is not reason that work cannot begin now.

Regaring `eye`, it is sometimes useful to get have all the memory allocated if you want to set some values later on.

**Comment by gajomi** [comment]
@andreasnoack I'm not quite sure I understood your comment about matrix reconstruction, but I agree that the current factorization printings are pretty confusing. If understand What Stefan is saying (and this might be somewhat orthogonal to the structure matrix printing issue, but bear with me) he would like to see the current behavior
```julia
>>A = rand(4,4)
4x4 Array{Float64,2}:
 0.0204562  0.198083  0.489023  0.00295775
 0.937587   0.151448  0.880757  0.919935  
 0.615964   0.950264  0.58058   0.184402  
 0.732427   0.726044  0.892177  0.317655 
>>F = qrfact(A)
Base.LinAlg.QRCompactWY{Float64,Array{Float64,2}}(4x4 Array{Float64,2}:
 -1.33991   -0.942716   -1.37835   -0.90217  
  0.689218   0.776919    0.167749  -0.392215 
  0.452794  -0.307224    0.468976   0.0521074
  0.538406  -0.0792158   0.174341  -0.103626 ,4x4 Array{Float64,2}:
 1.01527       -0.936176  -1.66063  0.0
 0.0            1.81709    1.13228  0.0
 0.0            0.0        1.941    0.0
 2.24097e-314   0.0        0.0      0.0)
```
evolve into something like
```julia
>>A = rand(4,4)
4x4 Array{Float64,2}:
 0.0204562  0.198083  0.489023  0.00295775
 0.937587   0.151448  0.880757  0.919935  
 0.615964   0.950264  0.58058   0.184402  
 0.732427   0.726044  0.892177  0.317655 
>>F = qrfact(A)
Base.LinAlg.QRCompactWY{Float64,Array{Float64,2}}
 0.0204562  0.198083  0.489023  0.00295775
 0.937587   0.151448  0.880757  0.919935  
 0.615964   0.950264  0.58058   0.184402  
 0.732427   0.726044  0.892177  0.317655 
```
where the elements are printed out identically in each case, but the header of `F`'s show method makes it clear that it is a factorization. This seems to me to be much nicer than the current behavior. But what I meant to suggest was something different still. I was imagining that the factorization display would focus rather on the factors so that one gets something like
```julia
>>A = rand(4,4)
4x4 Array{Float64,2}:
 0.0204562  0.198083  0.489023  0.00295775
 0.937587   0.151448  0.880757  0.919935  
 0.615964   0.950264  0.58058   0.184402  
 0.732427   0.726044  0.892177  0.317655 
>>F = qrfact(A)
Base.LinAlg.QRCompactWY{Float64,Array{Float64,2}}
Q: 4x4 OrthogonalMatrix{Float64,Array{Float64,2}}#OK, this isn't a yet a type
 -0.0152669   0.236435   0.913305  -0.33127      #but I am imagining it for the moment
 -0.699741   -0.654133   0.055439  -0.281778   
 -0.459707    0.665309  -0.35111   -0.471969   
 -0.546626    0.271241   0.198804   0.766881  
R: 4x4 UpperTriangular{Float64,Array{Float64,2}}:
 -1.33991  -0.942716  -1.37835   -0.90217  
            0.776919   0.167749  -0.392215 
                       0.468976   0.0521074
                                 -0.103626  
```
and where similarly the call `F[:R]` returns a triangular matrix with the pretty printing. You can imagine the generalization to SVD, Eigen, Cholesky and so on. I actually don't mean to say that this is superior to Stefan's idea. His proposal emphasizes the factorization object as equivalent in essential data to the original matrix, but useful for doing various computations efficiently by way of the factorization, which is certainly how they seem to be used in much of the Julia source. In contrast my proposal emphasizes the structure and values of the factor components (and in particular their characteristic zero valued regions), which is certainly also another way people might like to see factorizations.

**Comment by artkuo** [comment]
I was thinking more along the lines of @andreasnoack: Having a cleaner way of displaying structured matrices, the question is why the factorizations (other than Cholesky) don't return structured matrices. But now that I think about it, perhaps the issue is that LU and QR don't necessarily return square triangulars. If fed a square matrix, the factorization could indeed return structured triangulars. But when fed rectangular matrices, it would be a change in behavior to suddenly drop to dense "triangular" non-squares. The same goes for SVD where the sigma matrix takes the shape of the input matrix, whereas Cholesky returns structure because the input must be square.

If these issues did not matter, I would be in favor of the nice-looking R in QR, where Julia knows it is triangular and helps the user understand that by blanking the other triangle.

**Comment by andreasnoack** [comment]
That is exactly the reason why the factors of LU and QR are not structured. We have been discussin trapezoid structured matrices, but it is a bit of work to add new types and the advantages haven't been obvious relative to the work required.

**Comment by gajomi** [comment]
Thanks for the info. It was interesting to read through that issue and related ones ([#8240](https://github.com/JuliaLang/julia/issues/8240),[#8001](https://github.com/JuliaLang/julia/issues/8001)). I may want to get involved in some of this. But for the meantime, with respect to the printing issue here, I would advocate for skipping over factorization for the time being, at least until some of the work around them stabilizes.

**Comment by gajomi** [comment]
Behavior merged into master dc5c974df77129f3d129014e0acdb1997aeade8b.
